Analytical_Engine/Query/IsRectangle.cs

```diff
@@ -0,0 +1,113 @@
+﻿/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2020, the respective contributors. All rights reserved.
+ *
+ * Each contributor holds copyright over their respective contributions.
+ * The project versioning (Git) records all such contribution source information.
+ *                                           
+ *                                                                              
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.oM.Analytical.Elements;
+using BH.oM.Geometry;
+using BH.oM.Reflection.Attributes;
+using BH.Engine.Geometry;
+using BH.Engine.Reflection;
+using BH.Engine.Spatial;
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.ComponentModel;
+
+namespace BH.Engine.Analytical
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Determines whether a panel is a rectangle")]
+        [Input("panel", "The Panel to check if it is a rectangle")]
+        [Output("bool", "True for rectangular panels or false for non-rectangular panels")]
+        public static bool IsRectangle<TEdge, TOpening>(this IPanel<TEdge, TOpening> panel)
+            where TEdge : IEdge
+            where TOpening : IOpening<TEdge>
+        {
+            //Check if there are any openings
+            if (panel.Openings.Count != 0)
+                return false;
```

Comment:
So if a panel has openings, it is immediately not rectangular?

Reply:
Could see arguments both for and against this. Say for example this panel (which we deem valid in BHoM):

![image](https://user-images.githubusercontent.com/22005920/91280532-2379af80-e787-11ea-8cb7-cf60f95f0c64.png)

Should probably not be seen as a rectangle, but then, maybe this one should:

![image](https://user-images.githubusercontent.com/22005920/91280605-45733200-e787-11ea-9589-7e38d0aa6d97.png)


Might be worth adding a boolean argument, something like "acceptOpenings" or something like that?

Reply:
In that first picture, the host panel is still rectangular, that looks to be the green line to me? The opening having an edge coincident with the panel doesn't change the shape of the host panel. I imagine this is because in structures you wouldn't care if that was a window or not, to you that is a hole and is affecting the shape of the panel from a structural loading basis? Whereas in Environments, the existence of that window makes no difference to the host panel and thus the outline is still square.

Reply:
yeah, outline the same for both panels here.

And, yes, in structures an opening is basically stating "there is nothing here", same as not having a panel there at all, so would want to treat the top image as a non-rectangle. An U-shaped panel with only external edges would be equivalent for structural analysis/design.

But maybe the suggested boolean argument then could be a nice compromise? Also some potential to add an extension method to our opening classes, like `IsVoid()` or similar that could determine how the openings should be treated? Maybe both of these?

Reply:
But this is the Analytical Engine, not the Structural Engine. Structural Analysis is different from generic Analysis, so should a panel with an opening that could cause a U-Shaped panel be non-rectangular in the Analytical_Engine? I don't think it should, I feel that is an edge case which should be handled by an override in the Structural_Engine because, on the face of it, the panel itself is square. The openings have no bearing on that (in those examples)?

A boolean argument could resolve it, but I would still be wanting to question (and discuss or maybe reach agreement on) how that compromise doesn't compromise the Analytical Engine itself for future issues (with other disciplines maybe choosing different ways to categorise it)?

For example, another way round could be for Analytical Engine to say the first panel is rectangular, which Structural_Engine calls to determine, but then has an additional check. So Structures Engine can use this method to quickly rule out definite non-square panels, but then has additional checks for structural squareness?

Reply:
what You wrote, or, two separate methods?

`HasSquareOutline()` and `IsSolidSquare` or something like that? Adding it to Structure_Engine straight away kind of defeats the purpose of putting it in the Analytical_Engine in the first place IMO. Much more prefer a solution with a compromise in terms of arguments/method name.

Would say that simply adding an argument regarding if the geometry of all of the pieces of the panel should be accounted for or not to determine the squareness of a panel does not compromise the Analytical_Engine itself?

Reply:
It's more a case of how many future cases might we have to consider for other disciplines which leads me to want to look at the logic of squareness without the bits in the middle.

I would support `HasSquareOutline()` and `IsSolidSquare` - though the second might need a boolean argument to offer flexibility on the openings but then we're back to that argument again, which I don't think is needed. I would agree with this suggestion as is as a better alternative than the current proposal :smile:

Reply:
Ok I will make the change for `IsOutlineSquare()` and `IsOutlineRectangular` - perhaps we have a separate method for `ContainsOpenings` in the Analytical_Engine?